agent/immortalstreams/backedpipe/backed_pipe.go

```diff
@@ -0,0 +1,290 @@
+//
+// The writerSeqNum parameter is the local writer's current sequence number,
+// which should be sent to the remote side so it knows where to resume reading from.
+//
+// The returned readerSeqNum should be the remote side's current sequence number,
+// which indicates where the local reader should resume from.
+type ReconnectFunc func(ctx context.Context, writerSeqNum uint64) (conn io.ReadWriteCloser, readerSeqNum uint64, err error)
+
+// BackedPipe provides a reliable bidirectional byte stream over unreliable network connections.
+// It orchestrates a BackedReader and BackedWriter to provide transparent reconnection
+// and data replay capabilities.
+type BackedPipe struct {
+	ctx         context.Context
+	cancel      context.CancelFunc
+	mu          sync.RWMutex
+	reader      *BackedReader
+	writer      *BackedWriter
+	reconnectFn ReconnectFunc
+	conn        io.ReadWriteCloser
+	connected   bool
+	closed      bool
+
+	// Reconnection state
+	reconnecting bool
+
+	// Error channel for receiving connection errors from reader/writer
+	errorChan chan error
+
+	// singleflight group to dedupe concurrent ForceReconnect calls
+	sf singleflight.Group
+}
+
+// NewBackedPipe creates a new BackedPipe with default options and the specified reconnect function.
+// The pipe starts disconnected and must be connected using Connect().
+func NewBackedPipe(ctx context.Context, reconnectFn ReconnectFunc) *BackedPipe {
+	pipeCtx, cancel := context.WithCancel(ctx)
+
+	errorChan := make(chan error, 2) // Buffer for reader and writer errors
+	bp := &BackedPipe{
+		ctx:         pipeCtx,
+		cancel:      cancel,
+		reader:      NewBackedReader(),
+		writer:      NewBackedWriter(DefaultBufferSize, errorChan),
+		reconnectFn: reconnectFn,
+		errorChan:   errorChan,
+	}
+
+	// Set up error callback for reader only (writer uses error channel directly)
+	bp.reader.SetErrorCallback(func(err error) {
+		select {
+		case bp.errorChan <- err:
+		case <-bp.ctx.Done():
+		}
+	})
+
+	// Start error handler goroutine
+	go bp.handleErrors()
+
+	return bp
+}
+
+// Connect establishes the initial connection using the reconnect function.
+func (bp *BackedPipe) Connect() error {
+	bp.mu.Lock()
+	defer bp.mu.Unlock()
+
+	if bp.closed {
+		return ErrPipeClosed
+	}
+
+	if bp.connected {
+		return ErrPipeAlreadyConnected
+	}
+
+	// Use internal context for the actual reconnect operation to ensure
+	// Close() reliably cancels any in-flight attempt.
+	return bp.reconnectLocked()
+}
+
+// Read implements io.Reader by delegating to the BackedReader.
+func (bp *BackedPipe) Read(p []byte) (int, error) {
+	bp.mu.RLock()
+	reader := bp.reader
+	closed := bp.closed
+	bp.mu.RUnlock()
+
+	if closed {
+		return 0, io.EOF
+	}
+
+	return reader.Read(p)
+}
+
+// Write implements io.Writer by delegating to the BackedWriter.
+func (bp *BackedPipe) Write(p []byte) (int, error) {
+	bp.mu.RLock()
+	writer := bp.writer
+	closed := bp.closed
+	bp.mu.RUnlock()
+
+	if closed {
```

Comment:
Same deal as read --- the `writer` will return an error when it's closed, so you don't need to check the pipe. Also, writes don't return `EOF`.